mloop/neuralnet.py

```diff
@@ -638,43 +654,80 @@ def fit_neural_net(self, all_params, all_costs):
         all_params, all_costs = self._scale_params_and_cost_list(all_params, all_costs)
 
         if self.fit_hyperparameters:
-            # Every 20 runs, re-fit the hyperparameters.
+            # Re-fit the hyperparameters every runs_per_hyperparameter_fit runs.
+            runs_per_hyperparameter_fit = 100.0
             n_fits = len(all_params)
-            n_hyperfit = int(n_fits / 20.0)  # int() rounds down.
+            n_hyperfit = int(n_fits / runs_per_hyperparameter_fit)  # int() rounds down.
             if n_hyperfit > self.last_hyperfit:
+                self.log.info("Tuning regularization coefficient value.")
                 self.last_hyperfit = n_hyperfit
 
                 # Fit regularisation
 
-                # Split the data into training and cross validation
+                # May as well re-fit the cost and parameter scalers since a new
+                # net will be re-created from scratch. This keeps the scaled
+                # costs and parameters ~1.
+                self.scaler_samples = (all_params.copy(), all_costs.copy())
+                self._fit_scaler()
+
+                # Split the data into training and cross validation randomly
                 training_fraction = 0.9
-                split_index = int(training_fraction * len(all_params))
-                train_params = all_params[:split_index]
-                train_costs = all_costs[:split_index]
-                cv_params = all_params[split_index:]
-                cv_costs = all_costs[split_index:]
+                n_observations = len(all_costs)
+                split_index = int(training_fraction * n_observations)
+                indices = np.random.permutation(n_observations)
+                # Extract the training dataset.
+                train_indices = indices[:split_index]
+                train_params = all_params[train_indices]
+                train_costs = all_costs[train_indices]
+                # Extract the cross validation dataset.
+                cv_indices = indices[split_index:]
+                cv_params = all_params[cv_indices]
+                cv_costs = all_costs[cv_indices]
 
                 orig_cv_loss = self.net.cross_validation_loss(cv_params, cv_costs)
```

Comment:
Oh oops yep I forgot to delete that line from the previous version. Thanks for catching that!

Fortunately having it there doesn't change the behavior because `orig_cv_loss` is only used in the following line to set the initial value for `best_cv_loss`, but then the value of `best_cv_loss` is changed to `np.inf` on line 696 before it's used. Still though I'll issue another PR to clean that up.

Reply:
Oh right, I didn't see that we set it to `inf`.

Reply:
Fixed up in #96